### DISCUSSION TITLE
C++ hello world example should not use bad practice

### DISCUSSION THREAD

**Discussion by PeterSommerlad** [microsoft/vscode-cpptools#8029] [General]
using `using namespace std;` and outputting a new line with `std::endl` is bad practice in modern C++ and should not be shown to beginners who will think that is the only way to output a new line. it would be better to use

```
#include <iostream>

int main(){
    std::cout << "Hello, world!\n";
}
```

as a compile sanity check instead.

https://code.visualstudio.com/docs/languages/cpp#_add-hello-world-source-code

**Comment by michelleangela** [comment]
@PeterSommerlad 
Thank you for submitting this suggestion. Our documentation is managed by https://github.com/microsoft/vscode-docs/issues.

I can create this suggestion for you at https://github.com/microsoft/vscode-docs/issues if you'd like, or you're welcome to create another issue there too.

**Comment by michelleangela** [comment]
@PeterSommerlad
I created and referenced this suggestion at https://github.com/microsoft/vscode-docs/issues/4771

**Comment by sean-mcmanus** [comment]
What? std::endl is bad practice now? That is news to me. Where did you find that guideline? The std::endl was supposed account for \r\n on some platforms.

I know "using namespace std;" has always been not recommended.

**Reply by sreegowthamj** [reply]
https://stackoverflow.com/questions/4751972/endl-and-flushing-the-buffer

**Reply by sean-mcmanus** [reply]
Ah, yeah, I remember the buffer flushing now -- in a hello world example, you probably want it to flush the buffer, or else the user might not see any output when debugging, right? It's also useful when writing logging since if your app crashes, you want the logging to be flushed or it may not be up-to-date (i.e. it's more useful to have the logs right before the crash). Using endl is only bad when the flushing could cause a performance issue or when it's not otherwise beneficial. Maybe we could add more comments or something?

**Reply by PeterSommerlad** [reply]
Yeah, but destroying std::cout will flush the buffer anyway. And there is no guarantee the OS-internal buffers will actually flushed at program termination. So there is no win by adding `std::endl`. Also doing so in real code (i.e. with a  `std::ofstream` is a big pessimization, because it leads to OK round-trips making it slow, annihilating benefits through streambuf buffering. see "Sidestep System Calls" Pattern in https://hillside.net/europlop/HillsideEurope/Papers/EuroPLoP2002/2002_Sommerlad_PerformancePatterns.pdf

**Reply by sean-mcmanus** [reply]
When is std::cout destroyed? Another user just hit a buffer flush issue at https://github.com/microsoft/vscode-cpptools/discussions/8048 .

I'm not sure what you mean by there's no win by adding std::endl -- are you saying there are cases where it won't flush and you need to call std::flush or something?

For hello world type examples, I don't think performance nuances like this are important for new users to focus on. Do you know if any clang-tidy checks or another static code analysis tool will give a warning when std::endl is used?

**Reply by sean-mcmanus** [reply]
Okay, I found it at https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#slio50-avoid-endl

But for hello world, we'd want to make sure a user who has a breakpoint after the cout line will immediately see the output in the terminal or they may get confused (i.e. we occasionally get user issues like that, saying the debugger terminal output is broken, etc.).

**Reply by roalz** [reply]
> But for hello world, we'd want to make sure a user who has a breakpoint after the cout line will immediately see the output in the terminal or they may get confused (i.e. we occasionally get user issues like that, saying the debugger terminal output is broken, etc.).

If that is the only reason (questionable), I would suggest an explicit std::flush instead.

I.e.
https://stackoverflow.com/questions/22026751/c-force-stdcout-flush-print-to-screen

Example:
```
#include <iostream>

int main(){
    std::cout << "Hello, world!\n" << std::flush;
}
```
With a big, evident comment saying that you are flushing explicitely for the solely purpose of showing the result in the debugger terminal.
